ExitIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+    CarSensor testIS;
+    CarSensor testOS; 
+    ExitUI testUI;
+    Carpark testCarpark;
+    AdhocTicket  testAdhoc;
+    SeasonTicket testSeason;
+    IAdhocTicketDAO mockAdhocDAO;
+    ISeasonTicketDAO mockSeasonDAO;
+    ExitController testExit;
+    String seasonTicketId, cParkName;
+    UsageRecord mockRecord;
+    int cParkCap;
+    
+    @Before
+    public void setUp() {
+        cParkName = "test name";
+        cParkCap = 5;
+        mockAdhocDAO = mock(AdhocTicketDAO.class);
+        mockSeasonDAO = mock(SeasonTicketDAO.class);
+        mockRecord = mock(UsageRecord.class);
+        testCarpark = new Carpark(cParkName, cParkCap, mockAdhocDAO, mockSeasonDAO);
+        testGate = new Gate(0, 0);
+        testIS = new CarSensor("test inside", 100, 100);
+        testOS = new CarSensor("test outside", 200, 200);
+        testUI = mock(ExitUI.class);
+        testAdhoc = new AdhocTicket(cParkName, 1, "Atest");
+        testSeason = new SeasonTicket("test", cParkName, 1234, 4321);
+        testExit = new ExitController(testCarpark, testGate, testIS, testOS, testUI);
+    }
+    
+    @Test
+    public void initialisationTest() {
+        assertTrue(testCarpark instanceof Carpark);
+        assertTrue(testIS instanceof CarSensor);
+        assertTrue(testOS instanceof CarSensor);
+        assertTrue(testExit instanceof ExitController);
+    }
+    
+    @Test
+    public void testNormalFlowSeason() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");
+        
+        when(mockSeasonDAO.findTicketById(testSeason.getId())).thenReturn(testSeason);
+        testSeason.recordUsage(mockRecord);
+        testExit.ticketInserted(testSeason.getId());
+        assertEquals(testExit.getState(), "PROCESSED");
+        
+        testExit.ticketTaken();
+        verify(testUI).display("Take Processed Ticket");
+        assertEquals(testExit.getState(), "TAKEN");     
+        
+        testOS.setCarDetected(true);
+        testExit.carEventDetected(testOS.getId(), true);
+        assertEquals(testExit.getState(), "EXITING");
+        
+        testIS.setCarDetected(false);
+        testExit.carEventDetected(testIS.getId(), false);
+        assertEquals(testExit.getState(), "EXITED");
+    }
+    
+    @Test
+    public void testNormalFlowAdhoc() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");            //checks state transition and message display
+        
+        when(mockAdhocDAO.findTicketByBarcode("Atest")).thenReturn(testAdhoc);
+        testAdhoc.pay(1234, 1234);

```

Comment:
I don't think this is right.
1. tickets are not paid for at the exit pillar
2. how would the system get those numbers? If they are constants then while it may pass for this ticket it is bound to fail for every other ticket
3. the use case for Adhoc Ticket Normal Flow at this point says "System checks ticket paid or valid" 
you should be verifying that it has been paid, not paying it.
I think you need to use assertEquals(testAdhoc.getCharge(), 0) or assertNotEquals(testAdhoc.getPaidDateTime(), null)

Reply:
As before, the adhoc ticket testing was part of Aaron's area, so I took a few shortcuts here.
1.They aren't paid at the exit pillar, but here I'm setting the ticket to paid for the purposes of testing (this is normal flow, so the ticket is paid in this case).
2. They are just there as some values are needed to call the pay() method.
3. The verification is done in ExitController, as above, I just setup the necessary conditions.

Reply:
ok so having the ticket paid is part of the set up for this test, makes sense. Possibly putting your set up stuff at the top of the methods similar to how your class is laid out may be clearer, either way put comments explaining what is happening.

I believe you still are not proving that adhocTicket.isPaid() is being called and returning the correct value. Point 4 of the adhoc ticket normal flow use case says "4. System checks ticket paid or valid", it looks to me like you are not proving that is happening

Reply:
Fair enough, I see your point, I've added an assertEquals(testAdhoc.isPaid(), true) line there and commented on what it does.